10_rest/mohitnag/pkg/rest/rest.go

```diff
@@ -0,0 +1,130 @@
+package rest
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	"github.com/anz-bank/go-course/10_rest/mohitnag/pkg/puppy"
+	"github.com/anz-bank/go-course/10_rest/mohitnag/pkg/puppy/store"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"github.com/go-chi/render"
+)
+
+type Handler struct {
+	storer store.Storer
+	router chi.Router
+}
+
+func (rh *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rh.router.ServeHTTP(w, r)
+}
+
+func NewRestHandler(s store.Storer) http.Handler {
+	r := chi.NewRouter()
+	r.Use(middleware.RequestID)
+	r.Use(middleware.Logger)
+	r.Use(middleware.Recoverer)
+	r.Use(middleware.URLFormat)
+	r.Use(render.SetContentType(render.ContentTypeJSON))
+	rh := Handler{
+		storer: s,
+		router: r}
+	rh.handleRequests()
+	return &rh
+}
+
+func (rh *Handler) handleRequests() {
+	r := rh.router
+	r.Route("/api/puppy", func(r chi.Router) {
+		r.Route("/", func(r chi.Router) {
+			r.Post("/", rh.handlePostPuppy)
+			r.Put("/", rh.handleUpdatePuppy)
+		})
+		r.Route("/{id}", func(r chi.Router) {
+			r.Get("/", rh.handleGetPuppy)
+			r.Delete("/", rh.handleDeletePuppy)
+		})
+	})
+}
+
+func (rh *Handler) handlePostPuppy(w http.ResponseWriter, r *http.Request) {
+	var puppy puppy.Puppy
+	if err := render.DecodeJSON(r.Body, &puppy); err != nil {
+		renderDecodeJSONErrorResponse(w, r, err)
+		return
+	}
+	if err := rh.storer.CreatePuppy(puppy); err != nil {
+		renderErrorResponse(w, r, err)
+		return
+	}
+	w.WriteHeader(http.StatusCreated)
+}
+
+func (rh *Handler) handleGetPuppy(w http.ResponseWriter, r *http.Request) {
+	var puppy puppy.Puppy
+	ID := chi.URLParam(r, "id")
+	puppyID, err := strconv.Atoi(ID)
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+	}
+	puppy, err = rh.storer.ReadPuppy(uint32(puppyID))
+	if err != nil {
+		renderErrorResponse(w, r, err)
+		return
+	}
+	render.JSON(w, r, puppy)
+}
+
+func (rh *Handler) handleUpdatePuppy(w http.ResponseWriter, r *http.Request) {
+	var puppy puppy.Puppy
+	if err := render.DecodeJSON(r.Body, &puppy); err != nil {
+		renderDecodeJSONErrorResponse(w, r, err)
+		return
+	}
+	err := rh.storer.UpdatePuppy(puppy)
+	if err != nil {
+		renderErrorResponse(w, r, err)
+		return
+	}
+	render.JSON(w, r, puppy)
+}
+
+func (rh *Handler) handleDeletePuppy(w http.ResponseWriter, r *http.Request) {
+	ID := chi.URLParam(r, "id")
+	puppyID, err := strconv.Atoi(ID)
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
```

Comment:
You need a `return` after this line, otherwise you try to delete a puppy with a bad ID.

You might want to look at why your test case did not pick this up. I think you'll find that when you fix this, you won't have test coverage for the `err != nil` case below too, so you might need another test.

Reply:
Done